[Lim Kai Fang] iP

kflim opened the pull request:
### DukePad

> "Life is too complicated to not be orderly" - Martha Stewart ([source](https://dansilvestre.com/productivity-quotes/))

DukePad makes life easier by tracking your tasks. It is

- text-based
- _FAST_
- _FREE_ and _EASY_ to use

The steps are simple!
1. Download DukePad [here](url)
2. Add your tasks
3. Let DukePad remember all your tasks
4. Profit! 😉

Current features (Last updated 28 Aug):
- [X] Task management system
- [ ] Reminders (Coming soon)

For programmers familiar with Java, this is the `main` method.
```
public class Main {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

kflim committed:
Add greet message, echo command and exit message

kflim committed:
Add ability to store text by user and display them

kflim committed:
Add the ability to mark tasks as done

kflim committed:
Add support for tracking three types of tasks

kflim committed:
Add semi-automated testing of Duke

kflim committed:
Teach Duke to deal with errors

kflim committed:
Add support for deleting tasks from the list

kflim committed:
Save the tasks in the hard disk automatically

The tasks are only temporarily used during one execution of the program.

The program needs to be more reusable so that it can actually keep
track of all the tasks that have been made so far in the lifespan of
the program.

A storage for the list of tasks can be implemented to make the program
more reusable.

kflim committed:
Teach Duke how to understand dates and times

Dates entered are displayed as is.

The program should be able to display the dates in a more meaningful
way.

A way to understand different date formats can be implemented to make
the timings of the tasks more understandable.

kflim committed:
Merge branch 'branch-Level-7'

kflim committed:
Merge branch 'branch-Level-8'

kflim committed:
Introduce more OOP design to the program

The program tries to do everything in a few classes, particularly the
Duke class.

It is harder to modify the code as it all happens in one class.

Adding new classes will make modifying the code in the future easier.

kflim committed:
Add JUnit tests to test the behavior of the code

The code is presumably workable.

It is difficult to see corner cases that have not been accounted for.

Tests can help to reduce bugs in the code.

kflim committed:
Add JavaDoc comments to the code

Some of the public methods are documented for readability.

Some methods are not documented.

Documentation of methods can allow other developers to understand the
code better.

kflim committed:
Tweak the code to comply with a coding standard

 Code has some documentation.

 Code could use slight improvement.

 Tweaking the code could comply with the standard better.

kflim committed:
Add JavaDoc comments to the code

Code has some documentation

Some methods require further clarification.

More comments can help.

kflim committed:
Merge branch 'branch-A-JavaDoc'

kflim committed:
Tweak the code to comply with a coding standard

Code has some compliance.

Some import statements could be organized better

More changes can help.

kflim committed:
Merge branch 'branch-A-CodingStandard'

kflim committed:
Change packages organization

kflim committed:
Change organization of directories

kflim committed:
Merge branch 'branch-A-CodingStandard'

kflim committed:
Close scanner after use

kflim committed:
Add find task functionality

Program does not support find tasks by keyword

This makes finding tasks troublesome when the list becomes too big.

kflim committed:
Merge branch 'branch-Level-9'

kflim committed:
Merge remote-tracking branch 'origin/add-gradle-support'

kflim committed:
Use checkStyle to detect coding style violations

Code generally compiles well.

There are potential code violations according to the standard.

Using checkStyle can correct the violations.

kflim committed:
Add a GUI to Duke

Currently, Duke can only be interacted with through the CLI.

This is not very practical and is not very user-friendly.

A GUI is easier to interact with.

kflim committed:
Merge branch 'branch-JavaFx'

kflim committed:
Add changes to style

Code compiles fine.

Some parts of the code can be improved.

Some changes to style can be helpful.

kflim committed:
Merge branch 'branch-Level-10'

kflim committed:
Add some changes to code style

kflim committed:
Merge branch 'branch-Level-10'

kflim committed:
Configure checkstyle.xml

kflim committed:
Add header comments to methods

ErnestCuong reviewed (commented):
Looks good. There doesn't seem to be any coding violation.

andrewbaruch commented on src/main/java/bot/tasks/Event.java:
```diff
@@ -0,0 +1,92 @@
+        super(description);
+        String[] dateAndTime = timing.trim().split(" ", 2);
+
+        if (Pattern.matches(YMD_REGEX, dateAndTime[0])) {
```
The following kinda looks like an arrowhead! Since it seems the only difference is the format1, maybe you can try to extract it out?

kflim replied:
You're right! Will change this.

andrewbaruch commented on src/main/java/bot/utility/Duke.java:
```diff
@@ -0,0 +1,160 @@
+     * Runs the Duke chatBot
+     */
+    public void run() {
+        initialize();
```
Great use of SLAP!

kflim replied:
Thank you!

andrewbaruch commented on src/main/java/bot/utility/Duke.java:
```diff
@@ -0,0 +1,160 @@
+    /**
+     * Iteration 1:
+     * Creates a label with the specified text and adds it to the dialog container.
+     * @param text String containing text to add
```
Should there be an empty line somewhere here?

kflim replied:
Good catch! Will add a line

andrewbaruch commented on src/main/java/bot/tasks/Event.java:
```diff
@@ -0,0 +1,92 @@
+ * Represents a task that starts at a specific time and ends at a specific time.
+ */
+public class Event extends Task {
+    protected static final String YMD_DATE_FORMAT = "yyyy/MM/dd";
```
Would it be better if these constant is in Task so both Event and Deadline can access it without being copied?

kflim replied:
That would be better! Will change this

andrewbaruch commented on src/main/java/bot/utility/Parser.java:
```diff
@@ -0,0 +1,143 @@
+                task = tasks.get(idx - 1);
+                tasks.remove(idx - 1);
+                log.append("\n\t Noted. I've removed this task:\n\t\t").append(task)
+                        .append("\n\t Now you have ").append(tasks.size()).append(" tasks in the list.");
```
Would it be  better for readability if you use String format instead of multiple appends?

kflim replied:
Will look into this, thank you!

andrewbaruch reviewed (commented):
Hi Kai Fang!

Overall, it seems to be really good! There are a few changes which may be able to make your code more readable. Nonetheless, there seems to not be any major issue! Nice Work!

kflim committed:
Change code style

kflim committed:
Add Command abstractions

Parser currently does all of the work to interact with the logger and
tasklist.

In the long term, if there were new commands, the switch statement would
have to be very long. If there were specific changes to commands, it
would be difficult to edit.

Abstraction of the commands into different classes can allow for more
manageable code in the long term.

kflim committed:
Merge branch 'branch-Commands'

kflim committed:
Add header comments to Command classes

kflim committed:
Modify arrowhead code

Event has several arrowhead code.

Arrowhead code violates the coding standard.

Parts of the code are being abstracted into other methods.

Abstraction helps to make the code simpler and neater.

kflim committed:
Change Parser behaviour

Parsing does switch statement in the same try catch block as the error
checking.

The error checking should be separate from the actual switch cases.

Doing error checking first and then continue with the switch cases is
cleaner.

kflim committed:
Merge pull request #1 from kflim/A-CodeQuality

Change Parser behaviour

kflim committed:
Add minor changes.

Code is mostly functional.

Some comments may be not be complete.

Changing comments can improve code.

kflim committed:
Add minor changes to code.

Code has many comments.

Comments may not be structure correctly.

Minor adjustments can improve code.

kflim committed:
Add code changes

Code has some redundancies.

Code performance will be affected by the redundancies.

Removing redundacies will improve code.

kflim committed:
Change Parser behaviour

Parsing does switch statement in the same try-catch block as the error
checking.

The error checking should be separate from the actual switch cases.

Doing error checking first and then continue with the switch cases is
cleaner.

kflim committed:
Merge pull request #3 from kflim/branch-A-CodeQuality

Change Parser behaviour

kflim committed:
Merge branch 'branch-A-Assertions'

kflim committed:
Fix long methods

Some methods had more than 30 LoC.

If there were more similar details within the methods, the code would not
be maintained well.

Abstraction of some details into other methods can help to break up the
code into smaller functions so that the original method is much easier
to maintain.

kflim committed:
Improve the GUI and fix the creation flow of the GUI

The GUI is at the minimum level where an input is replied with a
generated response and there are basic features that the user can see.
The creation flow of the GUI is standard and all styled within the code itself.

It is not very visually pleasing and does not give unique responses to inputs.
The creation flow of the GUI can be changed to fit the abstraction of the bot
and styling is hard coded within the Window class.

Changing how the bot responds will help to increase interactivity with
the user. Abstracting some parts of the GUI creation to different classes
will make code more understandable and using .fxml files to style the
GUI will be simpler as well.

kflim committed:
Add a User Guide and fix some bugs

The readme.md file currently does not showcase how IntelliBot works.

First-time users will be confused with the lack of information.

Having a proper user guide will help first-time users understand how
IntelliBot works.

kflim committed:
Change README.md

kflim committed:
Set theme jekyll-theme-cayman

kflim committed:
Change README.md

kflim committed:
Merge branch 'master' of github.com:kflim/ip

kflim committed:
Set theme jekyll-theme-cayman

kflim committed:
Change README.md

kflim committed:
Merge branch 'master' of github.com:kflim/ip

kflim committed:
Set theme jekyll-theme-midnight

kflim committed:
Set theme jekyll-theme-cayman

kflim committed:
Fixed Ui.png name

kflim committed:
Merge branch 'master' of github.com:kflim/ip

kflim committed:
Fixed Ui.png image

kflim committed:
Release the product

kflim committed:
Fix long methods

kflim committed:
Fix import order

kflim committed:
Change indexing for DoneCommand

kflim committed:
Add screenshot
